src/mito2/src/compaction.rs

```diff
@@ -302,20 +279,81 @@ impl CompactionScheduler {
             for waiter in waiters {
                 waiter.send(Ok(0));
             }
-            return None;
+            self.region_status.remove(&region_id);
+            return Ok(());
         };
 
-        let task = CompactionTaskImpl {
+        // If specified to run compaction remotely, we schedule the compaction job remotely.
+        // It will fall back to local compaction if there is no remote job scheduler.
+        if current_version.options.compaction.remote_compaction() {
+            if let Some(remote_job_scheduler) = &self.plugins.get::<RemoteJobSchedulerRef>() {
+                let remote_compaction_job = CompactionJob {
+                    compaction_region: compaction_region.clone(),
+                    picker_output: picker_output.clone(),
+                    start_time,
+                };
+
+                let result = remote_job_scheduler
+                    .schedule(
+                        RemoteJob::CompactionJob(remote_compaction_job),
+                        Arc::new(DefaultNotifier {
+                            request_sender: request_sender.clone(),
+                            waiters: std::mem::take(&mut waiters),
+                        }),
+                    )
+                    .await;
```

Comment:
If the remote job scheduler returns an error, we will lose all `waiters`.

Reply:
If we follow the [original implementation](https://github.com/GreptimeTeam/greptimedb/blob/main/src/mito2/src/compaction/task.rs#L146-L173), maybe I need to add the following code:

```rust
impl DefaultNotifier {
    fn on_failure(&mut self, err: Arc<Error>, region_id: RegionId) {
        COMPACTION_FAILURE_COUNT.inc();
        for waiter in self.waiters.drain(..) {
            waiter.send(Err(err.clone()).context(CompactRegionSnafu {
                region_id,
            }));
        }
    }
}

#[async_trait::async_trait]
impl Notifier for DefaultNotifier {
    async fn notify(&mut self, result: RemoteJobResult) {
        match result {
            RemoteJobResult::CompactionJobResult(result) => {
                let notify = {
                    match result.region_edit {
                        Ok(edit) => BackgroundNotify::CompactionFinished(CompactionFinished {
                            region_id: result.region_id,
                            senders: std::mem::take(&mut self.waiters),
                            start_time: result.start_time,
                            edit,
                        }),
                        Err(err) => {
                            error!("Compaction failed for region {}: {:?}", result.region_id, err);
                            let err = Arc::new(err);
                            self.on_failure(err.clone(), result.region_id);
                            BackgroundNotify::CompactionFailed(CompactionFailed {
                                region_id: result.region_id,
                                err,
                            })
                        }
                    }
                };

                if let Err(e) = self
                    .request_sender
                    .send(WorkerRequest::Background {
                        region_id: result.region_id,
                        notify,
                    })
                    .await
                {
                    error!(
                        "Failed to notify compaction job status for region {}, error: {:?}",
                        result.region_id, e
                    );
                }
            }
        }
    }
}
```

Should we need to add the new type of error or we can put it in the new PR?

Reply:
The new PR has the following refactoring:

1. Add the `waiters` in `RemoteJobScheduler` error;
2. Move `waiters` from `DefaultNotifier` to `Job,` and it will be convenient to pass between multiple objects;